src/V3EmitCImp.cpp

```diff
@@ -145,7 +146,7 @@ class EmitCGatherDependencies final : VNVisitorConst {
     }
 
 public:
-    static const std::set<std::string> gather(AstCFunc* cfuncp) {
+    static const std::set<std::string> gather(AstCFunc* cfuncp) VL_MT_STABLE {
         const EmitCGatherDependencies visitor{cfuncp};
```

Comment:
EmitCGatherDependencies has many non-VL_MT_STABLE marked functions, so this should be failing the analyzer, and those functions need cleanup.  Or maybe there's another pre-pull needed before this can merge that has all those STABLEs.  Anyhow I'd like all the STABLEs markings right and committed before enabling the parallelism.

Reply:
@wsnyder Update to clang check annotations got merged and I've explained my reasoning about missing annotations above. I think this PR should be ready for review.

Reply:
I do understand your point about constructors - basically you are saying that the m_ methods are threadsafe in constructors which I agree with, but the problem is other calls using non-members e.g. globals are not.

If I add to main() a call to a non-stable function e.g. addCppFile:

    static void main(const AstNodeModule* modp, bool slow,
                     std::deque<AstCFile*>& cfilesr) VL_MT_STABLE {
        EmitCImp{modp, slow, cfilesr};
        v3Global.opt.addCppFile("FIXME");
    }

then I don't get a warning.  I think it's likely within the huge visitor there is something being called that isn't stable and we need to find/protect against that.   Of course, adding addCppFile to the constructor is also not flagged (same issue).

I appreciate this is dragging on, but I don't want any significant risk of hard to debug race cases once we enable threading.

Reply:
<Ignore last post - adding additional info>

Reply:
(Updated comment - ignore previous email version - I edited web version)

I do understand your point about constructors - basically you are saying that the m_ methods are threadsafe in constructors which I agree with, but the problem is other calls using non-members e.g. globals are not.

For example adding this to main() as you point out causes an attribute check failure as it should

    static void main(const AstNodeModule* modp, bool slow,
                     std::deque<AstCFile*>& cfilesr) VL_MT_STABLE {
        EmitCImp{modp, slow, cfilesr};
        v3Global.opt.addCppFile("FIXME");    // BAD - detected

the problem is adding it to the constructor does not.

     explicit EmitCImp(const AstNodeModule* modp, bool slow, std::deque<AstCFile*>& cfilesr)
        : m_fileModp{modp}
        , m_slow{slow}
        , m_cfilesr{cfilesr} {
        v3Global.opt.addCppFile("FIXME");   // BAD - not detected

The constructor can be called by multiple threads, and that means multiple calls to addCppFile, which will break things.

How can we detect these?

As you appreciate we need to make sure there are no subtle thread issues, as they will be nearly impossible to debug.

Reply:
I've fixed check for globals in constructor context in: https://github.com/verilator/verilator/pull/4171